### PR TITLE
Account app updates and cleanups

### PR DESCRIPTION
**Pull request by kschiffer**
#### Summary
This PR introduces fixes and updates to the current scaffold, which will prepare it for the upcoming PR on profile settings. It also contains refactors that will conform it to our current code style-guide, including the exclusive use of functional components, usage of arrow components. As such, the goal is also to reduce the diff of upcoming PRs.

#### Changes
- Convert all functions to arrow functions #3540 
- Convert all class components to functional components with hooks
- Remove context based usage of `env` (via `withEnv()` and `<EnvProvider />`)
- Update change password functionality
  - Use container class (to be used later-on in the authenticated view as well)
  - Use password validation rules from the IS (see #1067)
  - Add `updateUserPassword` rpc to JS SDK
- Add `requestPromiseMiddleware`
- Add redux primitives and logic for IS config
- Add SDK instance singleton module
- Some logic updates (using `createRequestLogic`)
- Optimize update password flow (e.g. redirect to login automatically)
- Fix routing
  - Handle root route via frontend (instead of backend redirect)
  - Use `?n=…` query parameter only for known authenticated routes 
  - Use not found page otherwise
- Replace usage of API tree with direct usage of the SDK instance
- Various small related fixes

#### Testing

I'm currently working on #3605 to be able to add some more tests, but would like this to be reviewed in the meantime already.

##### Regressions

All original OAuth functionality might experience regression with this.

#### Notes for Reviewers
Sorry for the big diff, it is mainly due to the functional component and arrow function updates. This could have been done successively as well but I really want the AA to be proper right from the start.

Again note that this PR targets the `feature/1422-account-app` feature branch, which I kept rebased on latest master.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/webui/account/views/overview/index.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+@connect(
+  state => ({
+    user: selectUser(state),
+  }),
+  {
+    logout,
+  },
+)
+export default class Overview extends React.PureComponent {
```
Comment (kschiffer) [inline pkg/webui/account/views/overview/index.js:37]: Note that this will be completely replaced in the next PR, which I why I did not update this one to a functional component.

File: pkg/webui/account/views/create-account/index.js [hunk 2]
```diff
@@ -91,151 +86,107 @@ const getSuccessMessage = state => {
...
+          {siteName}
+          <br />
+          <Message content={m.createANewAccount} component="strong" />
+        </h1>
+        <hr className={style.hRule} />
+        <Form
+          onSubmit={handleSubmit}
+          initialValues={initialValues}
+          error={error}
+          validationSchema={validationSchema}
```
Comment (bafonins) [inline pkg/webui/account/views/create-account/index.js:147]: Any reasons for not using https://github.com/TheThingsNetwork/lorawan-stack/blob/feature/1422-account-app-cleanups/pkg/webui/lib/create-password-validation-schema.js for this form?

File: pkg/webui/account/views/forgot-password/index.js [hunk 3]
```diff
@@ -58,81 +55,66 @@ const initialValues = { user_id: '' }
...
+          name="user_id"
+          component={Input}
+          autoFocus
+          required
+        />
+        <Form.Submit
+          component={SubmitButton}
+          message={m.send}
+          className={style.submitButton}
+          alwaysEnabled
```
Comment (bafonins) [inline pkg/webui/account/views/forgot-password/index.js]: Cant find the `alwaysEnabled` prop defined anywhere

File: pkg/webui/account/views/forgot-password/index.js [hunk 4]
```diff
@@ -58,81 +55,66 @@ const initialValues = { user_id: '' }
...
+        />
+        <Button.Link naked secondary message={sharedMessages.cancel} to="/login" />
```
Comment (bafonins) [inline pkg/webui/account/views/forgot-password/index.js]: 1. Navigate to `/console`
2. Land on the login page and go to the forgot password view
3. Submit or just navigate back
4. Login
5. Land on the account app overview instead of the Console

Not sure if we should preserve the initial users location, but is seems weird that we do that for registration and not for the forgot password view.

File: pkg/webui/account/views/front/index.js [hunk 5]
```diff
@@ -43,6 +47,11 @@ const FrontView = () => {
             <Route path="/validate" component={Validate} />
             <Route path="/code" component={Code} />
             <Redirect exact from="/" to="/login" />
+            {authRoutes.map(({ path, exact }) => (
+              <Route path={path} exact={exact} key={path}>
+                <Redirect to={`/login?n=${selectApplicationRootPath()}${location.pathname}`} />
```
Comment (bafonins) [inline pkg/webui/account/views/front/index.js:53]: What is the purpose of this redirect?

File: pkg/webui/account/views/front/index.js [hunk 6]
```diff
@@ -43,6 +47,11 @@ const FrontView = () => {
             <Route path="/validate" component={Validate} />
             <Route path="/code" component={Code} />
             <Redirect exact from="/" to="/login" />
+            {authRoutes.map(({ path, exact }) => (
```
Comment (bafonins) [inline pkg/webui/account/views/front/index.js:51]: Currently, there is only one route in `authRouter` which is for `/`. When the user navigates to `/oauth` these lines are never reached because of L49

File: pkg/webui/account/views/front/index.js [hunk 7]
```diff
@@ -26,10 +26,14 @@ import UpdatePassword from '@account/views/update-password'
 import FrontNotFound from '@account/views/front-not-found'
 import Validate from '@account/views/validate'
 import Code from '@account/views/code'
+import { authRoutes } from '@account/views/landing'
```
Comment (bafonins) [inline pkg/webui/account/views/front/index.js]: IMO, moving this to a separate file is better.

**Review by bafonins — COMMENTED** [review]
Submit section on forms looks a bit weird on tablets:

Screenshot 2020-12-27 at 16 33 35


Page content on the account overview completely disappears on phones:

Screenshot 2020-12-27 at 16 35 39

File: pkg/webui/account/views/front/index.js (hunk 5 above)
Comment (kschiffer) [reply pkg/webui/account/views/front/index.js:53]: The purpose is to redirect routes that require authentication back to the login page, with a query parameter that redirectes to the original route after login. E.g. `/profile-settings` (will be added later) would redirect back to the login page. All other routes that do not correspond to anything will show a 404.

Note that the `<Front />` handler only operates when the user is not authenticated.

File: pkg/webui/account/views/front/index.js (hunk 6 above)
Comment (kschiffer) [reply pkg/webui/account/views/front/index.js:51]: That's true. There will be more routes added with the next PRs which will make this setup purposeful. It's fine that `/` is handled by L49, since we don't need a redirect query parameter for it.

**Comment by kschiffer** [comment]
Thanks for the review. I've addressed all your comments.

File: pkg/webui/lib/create-password-validation-schema.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      password => hasMinLength(password, requirements.min_length),
+    )
+    .test(
+      'max-length',
+      { message: sharedMessages.validateTooLong, values: { max: requirements.max_length } },
+      password => hasMaxLength(password, requirements.max_length),
+    )
+    .test(
+      'min-special',
+      { message: sharedMessages.validateSpecial, values: { special: requirements.min_special } },
```
Comment (bafonins) [inline pkg/webui/lib/create-password-validation-schema.js:35]: `validateSpecial` is not defined in `shared-messages`

File: pkg/webui/lib/create-password-validation-schema.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      password => hasSpecial(password, requirements.min_special),
+    )
+    .test(
+      'min-upper',
+      { message: sharedMessages.validateUppercase, values: { upper: requirements.min_uppercase } },
```
Comment (bafonins) [inline pkg/webui/lib/create-password-validation-schema.js:40]: `validateUppercase` is not defined in `shared-messages`

File: pkg/webui/lib/create-password-validation-schema.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      password => hasUpper(password, requirements.min_uppercase),
+    )
+    .test(
+      'min-digit',
+      { message: sharedMessages.validateDigit, values: { digit: requirements.min_digits } },
```
Comment (bafonins) [inline pkg/webui/lib/create-password-validation-schema.js:45]: `validateDigit` is not defined in `shared-messages`

File: pkg/webui/account/views/validate/index.js [hunk 11]
```diff
@@ -45,70 +46,74 @@ const m = defineMessages({
...
+      } catch (error) {
+        if (isNotFoundError(error)) {
+          handleError(createFrontendError(m.tokenNotFoundTitle, m.tokenNotFoundMessage))
+        } else {
+          handleError(error)
+        }
       }
     }
-  }
+    makeRequest()
```
Comment (bafonins) [inline pkg/webui/account/views/validate/index.js:83]: Any reason for extracting the whole logic to a local function?

Comment (kschiffer) [reply pkg/webui/account/views/validate/index.js:83]: Because we need to run async logic there and it's not allowed to use an async function directly as argument for `useEffect()`, since the effect function must either return a cleanup function or nothing (but would then return a promise).

File: pkg/webui/account/store/middleware/user.js [hunk 12]
```diff
@@ -12,28 +12,38 @@
...
+  },
+})
+
+const getUserLogic = createRequestLogic({
+  type: user.GET_USER,
+  process: async ({ action, getState }) => {
+    const {
+      meta: { selector },
+    } = action
+    console.log(action)
```
Comment (bafonins) [inline pkg/webui/account/store/middleware/user.js]: Remove

**Comment by bafonins** [comment]
1. Go to `/oauth`
2. Fill in login/password and submit the form
3. Results in an endless loop without being logged in to the account app

Configuration I used for testing:

```
go run ./cmd/ttn-lw-stack start all \
--log.level debug \
--is.oauth.ui.js-file libs.bundle.js,account.js \
--console.ui.js-file libs.bundle.js,console.js \
--console.ui.assets-base-url http://localhost:8080/assets \
--console.ui.canonical-url http://localhost:8080/console \
--console.ui.is.enabled --console.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.as.enabled --console.ui.as.base-url http://localhost:8080/api/v3 \
--console.ui.ns.enabled --console.ui.ns.base-url http://localhost:8080/api/v3 \
--console.ui.js.enabled --console.ui.js.base-url http://localhost:8080/api/v3 \
--console.ui.gs.enabled --console.ui.gs.base-url http://localhost:8080/api/v3 \
--console.oauth.logout-url=http://localhost:8080/oauth/logout \
--console.oauth.authorize-url=http://localhost:8080/oauth/authorize \
--console.oauth.token-url=http://localhost:8080/oauth/token
```

**Comment by kschiffer** [comment]
Hmm that's odd. I cannot reproduce this. Does the same thing happen in incognito mode? It also doesn't seem to be reflected by the E2Es.

**Comment by bafonins** [comment]
> Does the same thing happen in incognito mode?

yes, same thing.

> It also doesn't seem to be reflected by the E2Es.

Try adding `cy.wait(10000)` at the end of `succeeds logging in when using valid credentials` test. Same issue happens when running cypress.

I guess we should revisit the way we assert page load

**Comment by kschiffer** [comment]
Got it now, you're missing `--is.oauth.ui.is.base-url http://localhost:8080/api/v3` in your configuration. Since this is a different host, the session cookie will not be sent to the IS making it return a 401.

Try like this:

```shell
go run ./cmd/ttn-lw-stack start \
--log.level debug \
--is.oauth.ui.js-file libs.bundle.js,account.js \
--console.ui.js-file libs.bundle.js,console.js \
--console.ui.assets-base-url http://localhost:8080/assets \
--console.ui.canonical-url http://localhost:8080/console \
--is.oauth.ui.canonical-url http://localhost:8080/oauth \
--is.oauth.ui.is.enabled --is.oauth.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.is.enabled --console.ui.is.base-url http://localhost:8080/api/v3 \
--console.ui.as.enabled --console.ui.as.base-url http://localhost:8080/api/v3 \
--console.ui.ns.enabled --console.ui.ns.base-url http://localhost:8080/api/v3 \
--console.ui.js.enabled --console.ui.js.base-url http://localhost:8080/api/v3 \
--console.ui.gs.enabled --console.ui.gs.base-url http://localhost:8080/api/v3 \
--console.oauth.logout-url=http://localhost:8080/oauth/logout \
--console.oauth.authorize-url=http://localhost:8080/oauth/authorize \
--console.oauth.token-url=http://localhost:8080/oauth/token
```

These are the additional ENVs I use on top [of the usual dev config setup](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.10/DEVELOPMENT.md#development-configuration):
```
export TTN_LW_IS_OAUTH_UI_JS_FILE="libs.bundle.js account.js"
export TTN_LW_IS_OAUTH_UI_IS_BASE_URL="http://localhost:8080/api/v3"
```

File: cmd/internal/shared/identityserver/config.go [hunk 13]
```diff
@@ -28,15 +28,17 @@ import (
...
 				SiteName:      "The Things Stack for LoRaWAN",
+				Title:         "Account",
 				Language:      "en",
 				CanonicalURL:  shared.DefaultOAuthPublicURL,
 				AssetsBaseURL: shared.DefaultAssetsBaseURL,
 				IconPrefix:    "oauth-",
-				CSSFiles:      []string{"oauth.css"},
-				JSFiles:       []string{"oauth.js"},
+				CSSFiles:      []string{"account.css"},
+				JSFiles:       []string{"account.js"},
```
Comment (htdvisser) [inline cmd/internal/shared/identityserver/config.go]: This is a breaking change for deployments that use custom JS/CSS files (which I don't think is very common). In any case, make sure to mention in CHANGELOG that deployments need to change their config if they customized these.

File: pkg/identityserver/identityserver.go [hunk 14]
```diff
@@ -106,10 +108,18 @@ func New(c *component.Component, config *Config) (is *IdentityServer, err error)
 		store.ClientStore
 		store.OAuthStore
 	}{
-		UserStore:        store.GetUserStore(is.db),
-		UserSessionStore: store.GetUserSessionStore(is.db),
 		ClientStore:      store.GetClientStore(is.db),
 		OAuthStore:       store.GetOAuthStore(is.db),
+		UserStore:        store.GetUserStore(is.db),
+		UserSessionStore: store.GetUserSessionStore(is.db),
```
Comment (htdvisser) [inline pkg/identityserver/identityserver.go]: Unnecessary diff

**Review by htdvisser — COMMENTED** [review]
Please update CODEOWNERS so that there are code owners on `pkg/account`

File: pkg/identityserver/identityserver.go (hunk 14 above)
Comment (htdvisser) [reply pkg/identityserver/identityserver.go]: I'm still seeing the diff in the latest version of the PR

**Review by htdvisser — CHANGES_REQUESTED** [review]
This still doesn't have CODEOWNERS on `pkg/account` as I requested in an earlier review. Please address previous comments before re-requesting review.

**Comment by kschiffer** [comment]
> This still doesn't have CODEOWNERS on `pkg/account` as I requested in an earlier review. Please address previous comments before re-requesting review.

Sorry, missed a push there.

File: CODEOWNERS [hunk 15]
```diff
@@ -58,6 +58,7 @@
 /pkg/identityserver @htdvisser @bafonins @kschiffer
 /pkg/email @htdvisser @bafonins @kschiffer
 /pkg/oauth @htdvisser @bafonins @kschiffer
+/pkg/account @htdvisser @bafonins @kschiffer
```
Comment (johanstokking) [inline CODEOWNERS:61]: This is my only code owned file and I'm fine with this if this was discussed before.

Maybe loop in @asmulko or @mjamescompton instead of @htdvisser to reduce his scope

**Review by johanstokking — APPROVED** [review]
Please let me know if you want me to review something specifically, and if so, what

File: CODEOWNERS (hunk 15 above)
Comment (htdvisser) [reply CODEOWNERS:61]: We should indeed take a better look at the packages that are code-owned by me. In this case it's fine to add me me. (1) It's pretty much all code that was moved from `pkg/oauth`, so I already owned it. (2) It's pretty small and (3) I don't expect it to change often.

**Comment by kschiffer** [comment]
> Please let me know if you want me to review something specifically, and if so, what

I don't think there's anything else for you to review in this.

File: pkg/webui/account/store/middleware/user.js [hunk 16]
```diff
@@ -12,28 +12,37 @@
...
+})
+
+const getUserLogic = createRequestLogic({
+  type: user.GET_USER,
+  process: async ({ action, getState }) => {
+    const {
+      meta: { selector },
+    } = action
+    const userId =
+      'payload' in action && action.payload.id ? action.payload.id : selectUserId(getState())
```
Comment (bafonins) [inline pkg/webui/account/store/middleware/user.js:40]: it seems that id` is required here. any use case when it can be missing?

File: pkg/webui/account/store/reducers/identity-server.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { handleActions } from 'redux-actions'
+
+import { GET_IS_CONFIGURATION_SUCCESS } from '@console/store/actions/identity-server'
```
Comment (bafonins) [inline pkg/webui/account/store/reducers/identity-server.js]: I dont think we should reference console and account apps. Some repetition wont hurt until we implement some proper api to share store related functionality between apps.

Comment (bafonins) [reply pkg/webui/account/store/reducers/identity-server.js]: Actually, there are is actions defined in `@account/store/actions/identity-server`

File: pkg/webui/account/views/create-account/index.js [hunk 18]
```diff
@@ -91,151 +87,124 @@ const getSuccessMessage = state => {
...
-      })
-    } catch (error) {
-      this.setState({
-        error: error.response.data,
-      })
-    } finally {
-      setSubmitting(false)
-    }
+  const [error, setError] = useState(undefined)
+  const passwordRequirements = useSelector(selectPasswordRequirements)
```
Comment (bafonins) [inline pkg/webui/account/views/create-account/index.js:97]: You can get the is configuration directly from `useRequest` on L87

File: pkg/webui/account/views/login/index.js [hunk 19]
```diff
@@ -59,112 +57,107 @@ const validationSchema = Yup.object().shape({
...
-    return next.split('?')[0]
+  if (location.state && location.state.info) {
+    info = location.state.info
+  } else if (!next || (next !== appRoot && !Boolean(error))) {
+    info = m.loginToContinue
   }
 
-  return next
+  return (
+    <React.Fragment>
```
Comment (bafonins) [inline pkg/webui/account/views/login/index.js]: fragment not needed

File: pkg/webui/account/views/login/index.js [hunk 20]
```diff
@@ -59,112 +57,107 @@ const validationSchema = Yup.object().shape({
...
+            <Form.Submit
+              component={SubmitButton}
+              message={sharedMessages.login}
+              className={style.submitButton}
+            />
+            <Button.Link
+              to={`/register${location.search}`}
+              secondary
+              message={m.createAccount}
+              className={style.registerButton}
```
Comment (bafonins) [inline pkg/webui/account/views/login/index.js]: `register-button` class does not exist

File: pkg/webui/account/views/front/index.js [hunk 21]
```diff
@@ -27,9 +29,12 @@ import FrontNotFound from '@account/views/front-not-found'
...
+import { selectApplicationRootPath } from '@ttn-lw/lib/selectors/env'
+import PropTypes from '@ttn-lw/lib/prop-types'
+
 import style from './front.styl'
 
-const FrontView = () => {
+const FrontView = ({ location }) => {
   return (
     <div className={style.container}>
       <section className={style.content}>
```
Comment (bafonins) [inline pkg/webui/account/views/front/index.js:40]: On the next line `.header` class does not exist.

File: pkg/webui/console/components/user-data-form/index.js [hunk 22]
```diff
@@ -79,53 +74,14 @@ const validationSchema = Yup.object().shape({
...
 @injectIntl
 class UserForm extends React.Component {
   constructor(props) {
     super(props)
 
     const { update, passwordRequirements } = props
     this.validationSchema = update
-      ? validationSchema
+      ? baseValidationSchema
       : createPasswordValidationSchema(passwordRequirements)
```
Comment (bafonins) [inline pkg/webui/console/components/user-data-form/index.js]: We need to merge password validation with validation rules for other fields as well. With this implementation:
1. Form values are no properly casted
2. Multiple fields are missing validation

File: pkg/webui/lib/components/full-view-error/error.js [hunk 23]
```diff
@@ -33,10 +32,13 @@ import {
...
 import statusCodeMessages from '@ttn-lw/lib/errors/status-code-messages'
 import PropTypes from '@ttn-lw/lib/prop-types'
+import { selectApplicationRootPath } from '@ttn-lw/lib/selectors/env'
 
 import style from './error.styl'
 
-const FullViewErrorInner = function({ error, env }) {
+const appRoot = selectApplicationRootPath()
+
+const FullViewErrorInner = function({ error }) {
```
Comment (bafonins) [inline pkg/webui/lib/components/full-view-error/error.js]:
```suggestion
const FullViewErrorInner = ({ error }) => {
```

File: pkg/webui/lib/components/full-view-error/error.js [hunk 24]
```diff
@@ -88,22 +90,19 @@ const FullViewErrorInner = function({ error, env }) {
   )
 }
 
-const FullViewErrorInnerWithEnv = withEnv(FullViewErrorInner)
-
 const FullViewError = function({ error, header }) {
```
Comment (bafonins) [inline pkg/webui/lib/components/full-view-error/error.js]:
```suggestion
const FullViewError = ({ error, header }) => {
```

File: pkg/webui/account.js [hunk 25]
```diff
@@ -15,40 +15,35 @@
...
-const rootElement = document.getElementById('app')
+if (selectSentryDsnConfig()) {
+  Sentry.init(sentryConfig)
+}
 
 const render = () => {
   const App = require('./account/views/app').default
 
   DOM.render(
-    <EnvProvider env={env}>
```
Comment (bafonins) [inline pkg/webui/account.js:42]: We still have base components that rely on `withEnv` hoc, e.g. `<BaseAnchorLink />` and `<DocLink />` that can be used in the account app. We should keep the `<EnvProvider />` as long as it is used in `@ttn-lw/conponents/`.

**Review by bafonins — CHANGES_REQUESTED** [review]

**Review by htdvisser — APPROVED** [review]

File: pkg/webui/account/store/middleware/user.js (hunk 16 above)
Comment (kschiffer) [reply pkg/webui/account/store/middleware/user.js:40]: Once the user is logged in, we can omit the user id and get it from the store. This does not work initially during the init logic. Since after the initialization the user is stored, there should be no situation where the user id is missing.

File: pkg/webui/account/store/reducers/identity-server.js (hunk 17 above)
Comment (kschiffer) [reply pkg/webui/account/store/reducers/identity-server.js]: Yup. Forgot to rename the import.

File: pkg/webui/account/views/create-account/index.js (hunk 18 above)
Comment (kschiffer) [reply pkg/webui/account/views/create-account/index.js:97]: I can but it would need an extra selector to extract the password requirements from the is config object (and not from the root of the store). I don't think it's worth the hassle.

File: pkg/webui/account.js (hunk 25 above)
Comment (kschiffer) [reply pkg/webui/account.js:42]: Searching in `pkg/webui/components` reveals that `<BaseAnchorLink />` and `<DocLink />` seem to be the only one's that still use `withEnv`. I'll add a commit to replace this with env selectors.

**Comment by kschiffer** [comment]
@bafonins thanks for the review. I've addressed all your comments and also did another rebase on latest `v3.11`. I had to add [some additional commits](https://github.com/TheThingsNetwork/lorawan-stack/pull/3613/commits/a817d1de1bc44cc67d668b89ece9f5b91db04f18) to properly merge online status functionality, which was introduced via #3629.

**Review by bafonins — APPROVED** [review]
Manually went and tested through various pages of the account app. Didnt review the code completely, only changes since the last review.
